### PR TITLE
more efficient and correct clear in Slf4jLogger

### PR DESCRIPTION
**Pull request by patriknw**
When reading https://github.com/akka/akka/issues/28077 I saw that we can set and clear the MDC values in a more efficient way in Slf4jLogger.

### CODE REVIEW

File: akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala [hunk 1]
```diff
@@ -104,18 +104,15 @@ class Slf4jLogger extends Actor with SLF4JLogging with RequiresMessageQueue[Logg
 
   @inline
   final def withMdc(logSource: String, logEvent: LogEvent)(logStatement: => Unit): Unit = {
-    MDC.put(mdcAkkaSourceAttributeName, logSource)
-    MDC.put(mdcThreadAttributeName, logEvent.thread.getName)
-    MDC.put(mdcAkkaTimestamp, formatTimestamp(logEvent.timestamp))
-    MDC.put(mdcActorSystemAttributeName, actorSystemName)
-    logEvent.mdc.foreach { case (k, v) => MDC.put(k, String.valueOf(v)) }
+    val mdcAdapter = MDC.getMDCAdapter
+    mdcAdapter.put(mdcAkkaSourceAttributeName, logSource)
```
Comment (patriknw) [inline akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala]: avoid accessing the ThreadLocal for each value

File: akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala [hunk 2]
```diff
@@ -104,18 +104,15 @@ class Slf4jLogger extends Actor with SLF4JLogging with RequiresMessageQueue[Logg
...
+    mdcAdapter.put(mdcActorSystemAttributeName, actorSystemName)
+    logEvent.mdc.foreach { case (k, v) => mdcAdapter.put(k, String.valueOf(v)) }
     try logStatement
     finally {
-      MDC.remove(mdcAkkaSourceAttributeName)
-      MDC.remove(mdcThreadAttributeName)
-      MDC.remove(mdcAkkaTimestamp)
-      MDC.remove(mdcActorSystemAttributeName)
-      logEvent.mdc.keys.foreach(k => MDC.remove(k))
+      MDC.clear()
```
Comment (patriknw) [inline akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala]: If there were some other MDC values in the ThreadLocal that would be invalid usage of MDC so should be fine to clear the whole thing. Also, I think removing all entries still leaves something in the Thread so a small leak to not clear.

**Comment by atooni** [comment]
Hi Patrick, 

thanks for your answer - and of course you are correct. Meanwhile I have reread the documentation; I was mislead by the comment in the docs that an insert into the MDC should happen rather rarely. So I was under the (wrong) assumption that it is populated once and then left for others to use. 

In our application I have now found a way that works. At the end of the day we just wanted to enrich log events that are being sent to splunk with more or less statis data to indicate where the message came from. 

From my perspective the issue can be closed.

**Comment by patriknw** [comment]
Thanks

File: akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala (hunk 1 above)
Comment (johanandren) [reply akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala]: Feels like a dejavu but if I look at hte Logback impl it accesses the thread local for each `put` as well, so there's no difference?

Comment (patriknw) [reply akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala]: you are right, misunderstand of me. I thought the MDCAdapter was *the* ThreadLocal value

an alternative would be `setContextMap`, but that creates a new copy-Map so might not be better

File: akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala (hunk 2 above)
Comment (patriknw) [reply akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala]: @johanandren what do you think I should do with this PR? I still think the MDC.clear() at the end is more correct and more efficient. Using the mdcAdapter for the put doesn't help much so I can revert that part?

Comment (johanandren) [reply akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala]: Sounds good to me.

Comment (patriknw) [reply akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala]: done, also changed other places that used mdcAdapter.put

**Review by johanandren — APPROVED** [review]

File: akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala (hunk 2 above)
Comment (chbatey) [reply akka-slf4j/src/main/scala/akka/event/slf4j/Slf4jLogger.scala]: :+1:

**Review by chbatey — APPROVED** [review]
LGTM

**Pull request merged**